src/mlpack/tests/decision_tree_test.cpp

```diff
@@ -1213,4 +1217,38 @@ BOOST_AUTO_TEST_CASE(DecisionTreeCategoricalTrainReturnEntropy)
   BOOST_REQUIRE_EQUAL(std::isfinite(entropy), true);
 }
 
+/**
+ * Make sure different Maximum Depth gives different number of childern.
+ */
+BOOST_AUTO_TEST_CASE(DifferentMaximumDepthTest)
+{
+  arma::mat dataset(10, 100, arma::fill::randu);
+  arma::Row<size_t> labels(100);
+  for (size_t i = 0; i < 50; ++i)
+  {
+    dataset(3, i) = 0.0;
+    labels[i] = 0;
+  }
+  for (size_t i = 50; i < 100; ++i)
+  {
+    dataset(3, i) = 1.0;
+    labels[i] = 1;
+  }
+
+  DecisionTree<> d(dataset, labels, 2, 10, 1e-7, 1);
+
+  DecisionTree<> d1(dataset, labels, 2, 10, 1e-7, 2);
+
+  DecisionTree<> d2(dataset, labels, 2, 10, 1e-7, 0);
+
+  // Now require that we have zero children.
+  BOOST_REQUIRE_EQUAL(d.NumChildren(), 0);
+
+  // Now require that we have zero children.
+  BOOST_REQUIRE_GT(d1.NumChildren(), 0);
+
+  // Now require that we have zero children.
+  BOOST_REQUIRE_GT(d2.NumChildren(), 0);
```

Comment:
The comment seem inaccurate here.  You might also consider checking deeper in the tree... so for `d1` you might check...

```
BOOST_REQUIRE_EQUAL(d1.NumChildren(), 2);
BOOST_REQUIRE_EQUAL(d1.Child(0).NumChildren(), 0);
BOOST_REQUIRE_EQUAL(d1.Child(0).NumChildren(), 0);
```

and then for `d2`,

```
BOOST_REQUIRE_EQUAL(d2.NumChildren(), 2);
BOOST_REQUIRE_EQUAL(d2.Child(0).NumChildren(), 2);
BOOST_REQUIRE_EQUAL(d2.Child(1).NumChildren(), 2);
```

Reply:
I have made the improvement as suggested.